tools/ci/marshalcheck/marshalcheck.go

```diff
@@ -0,0 +1,246 @@
+                namedStruct := findStructByName(pass, elemIdent.Name)
+                if namedStruct != nil {
+                    analyzeStructFields(pass, namedStruct, elemIdent.Name, fieldPath+"[]", isMergeable)
+                }
+            }
+        }
+
+        // Process the field normally
+        if isMergeable {
+            if !isPointer(field) {
+                pass.Reportf(field.Pos(), "mergeable field %s must be a pointer", fieldPath)
+            }
+            if !hasOmitEmptyTag(field) {
+                pass.Reportf(field.Pos(), "mergeable field %s must have 'omitempty' in JSON tag", fieldPath)
+            }
+            if hasDefaultAnnotation(pass, field) {
+                pass.Reportf(field.Pos(), "mergeable field %s must not have '+kubebuilder:default' annotation", fieldPath)
+            }
+        } else {
+            _, isValid := determineNonMergeableCategory(pass, field)
+
+            if !isValid {
+                pass.Reportf(field.Pos(), "field %s does not match any allowed non-mergeable category", fieldPath)
+            }
+        }
+    }
+}
+
+
+func findStructByName(pass *analysis.Pass, structName string) *ast.StructType {
+    var foundStruct *ast.StructType
+
+    for _, file := range pass.Files {
+        ast.Inspect(file, func(n ast.Node) bool {
+            typeSpec, ok := n.(*ast.TypeSpec)
+            if !ok {
+                return true
+            }
+            if typeSpec.Name.Name != structName {
+                return true
+            }
+            if structType, ok := typeSpec.Type.(*ast.StructType); ok {
+                foundStruct = structType // Save the struct in the variable
+                return false             // Stop further inspection
+            }
+            return true
+        })
+        if foundStruct != nil {
+            return foundStruct
+        }
+    }
+    return nil
+}
+
+func isPointerToSlice(field *ast.Field) bool {
+    if ptr, ok := field.Type.(*ast.StarExpr); ok {
+        _, isArray := ptr.X.(*ast.ArrayType)
+        return isArray
+    }
+    return false
+}
+
+func determineNonMergeableCategory(pass *analysis.Pass, field *ast.Field) (string, bool) {
+    hasDefault := hasDefaultAnnotation(pass, field)
+    hasOmitEmpty := hasOmitEmptyTag(field)
+    isPtr := isPointer(field)
+
+    if isPtr && hasOmitEmpty && !hasDefault {
+        return "optional_without_default", true
+    }
+    if !isPtr && hasDefault && !hasOmitEmpty {
+        return "optional_with_default", true
+    }
+    if !isPtr && !hasDefault && !hasOmitEmpty {
+        return "required", true
+    }
+    return "", false
+}
+
+func hasDefaultAnnotation(pass *analysis.Pass, field *ast.Field) bool {
+    if pass.Files == nil {
+        return false
+    }
+
+    fieldPos := field.Pos() // The position of the field in the source file
+    for _, commentGroup := range pass.Files {
+        for _, comment := range commentGroup.Comments {
```

Comment:
There is a `Doc` field in the [Field](https://pkg.go.dev/go/ast#Field) type. Can we use it? 

Also I don't really understand how `hasDefaultAnnotation` function works, how it prevents annotation from fields above not being taken into account, i.e:
```go
type MyStruct struct {
    // +kubebuilder:default:a
    A TypeA
    B TypeB
}
```
when you call `hasDefaultAnnotation` for `B`, is it going to associate `A`'s default annotation since it's Pos is less than `B`'s Pos?

Reply:
You're 100% right, `Doc` is the correct field.